src/equations/2d/ideal_glm_mhd.jl

```diff
@@ -453,6 +453,75 @@ function noncons_interface_flux(u_left, u_right, orientation, equation::IdealGlm
   return SVector(0, noncons2, noncons3, noncons4, noncons5, noncons6, noncons7, noncons8, noncons9)
 end
 
+# strong form of nonconservative flux on a side, e.g., the Powell term
+#     phi^L 1/2 (B^L+B^R) normal - phi^L B^L normal = phi^L 1/2 (B^R-B^L) normal
+# OBS! 1) this is non-unique along an interface! normal direction is super important
+function noncons_interface_flux_whole(u_left, u_right, orientation, equation::IdealGlmMhdEquations2D)
```

Comment:
so for the shock capturing form we need an alternative non-conservative term function to compute the `left` and `right` fluxes. We could probably streamline this with a flag of some kind to avoid code repetition because only two entries are altered for each orientation.

Reply:
Yes, we need two new non-conservative surface fluxes for the shock capturing. There is indeed some repeated code, but I decided to do it that way in order not to affect performance. As a matter of fact, an extra conditional evaluation in the non-conservative surface flux might affect simulations that do not use shock capturing. Any ideas of how to do the streamlining?

Reply:
You could try to add an additional argument and rely on constant propagation (as we do for the direction). Then, you could update the lines
```julia
  if orientation == 1 # x-direction
    v_normal = v1_ll
    B_normal = B1_rr - B1_ll
  else # y-direction
    v_normal = v2_ll
    B_normal = B2_rr - B2_ll
  end
```
to also set some required variables given the new argument. Then, you should benchmark and check the performance.

Reply:
I added a new argument to the function `noncons_interface_flux` and unified all the different non-conservative terms there. Then I performed a benchmark analysis.

I tested the rotor with `@benchmark Trixi.run("examples/2d/parameters_rotor.toml",t_end=0.05) evals=20` and obtained:

1. With the new implementation:
```
BenchmarkTools.Trial: 
  memory estimate:  223.61 MiB
  allocs estimate:  239901
  --------------
  minimum time:     4.280 s (0.66% GC)
  median time:      4.280 s (0.66% GC)
  mean time:        4.280 s (0.66% GC)
  maximum time:     4.280 s (0.66% GC)
  --------------
  samples:          1
  evals/sample:     20
```
2. With the old implementation:
```
BenchmarkTools.Trial: 
  memory estimate:  223.61 MiB
  allocs estimate:  239901
  --------------
  minimum time:     4.282 s (0.69% GC)
  median time:      4.282 s (0.69% GC)
  mean time:        4.282 s (0.69% GC)
  maximum time:     4.282 s (0.69% GC)
  --------------
  samples:          1
  evals/sample:     20
```

I also tested the Alfvén wave test case with `@benchmark Trixi.run("examples/2d/parameters_alfven_wave.toml") evals=20` and obtained:

1. With the new implementation:

```
BenchmarkTools.Trial: 
  memory estimate:  42.03 MiB
  allocs estimate:  574968
  --------------
  minimum time:     2.208 s (0.17% GC)
  median time:      2.208 s (0.17% GC)
  mean time:        2.208 s (0.17% GC)
  maximum time:     2.208 s (0.17% GC)
  --------------
  samples:          1
  evals/sample:     20
```
2. With the old implementation:
```
BenchmarkTools.Trial: 
  memory estimate:  42.03 MiB
  allocs estimate:  574968
  --------------
  minimum time:     2.205 s (0.18% GC)
  median time:      2.205 s (0.18% GC)
  mean time:        2.205 s (0.18% GC)
  maximum time:     2.205 s (0.18% GC)
  --------------
  samples:          1
  evals/sample:     20
```

So, I suppose that we want to keep the new implementation, as it has less repeated code and the performance is not significantly affected, right?

P.S. In case you are interested, the new routine is:

```
"""
    noncons_interface_flux(u_left, u_right, orientation, mode, equation::IdealGlmMhdEquations2D)
    
Strong form of nonconservative flux on a side (Powell and GLM terms)
     phi^L 1/2 (B^L+B^R) normal - phi^L B^L normal = phi^L 1/2 (B^R-B^L) normal
OBS! 1) The non-conservative interface flux depends on the discretization. Following "modes" are available:
        * 1: 'weak' formulation of split DG already includes the contribution -1/2(phi^L B^L normal)
             so this mode only adds 1/2(phi^L B^R normal)
             ... analogously for the Galilean nonconservative term
        * 2: This mode adds the whole non-conservative term: phi^L 1/2 (B^R-B^L)
        * 3: This mode adds the split-form DG volume integral contribution: This is equivalent to (2)-(1)
             - 1/2(phi^L B^L)
     2) this is non-unique along an interface! normal direction is super important
"""
function noncons_interface_flux(u_left, u_right, orientation, mode, equation::IdealGlmMhdEquations2D)
  rho_ll, rho_v1_ll, rho_v2_ll, rho_v3_ll, _, B1_ll, B2_ll, B3_ll, psi_ll = u_left
  _, _, _, _, _, B1_rr, B2_rr, _, psi_rr = u_right

  # extract velocites from the left
  v1_ll  = rho_v1_ll / rho_ll
  v2_ll  = rho_v2_ll / rho_ll
  v3_ll  = rho_v3_ll / rho_ll
  v_dot_B_ll = v1_ll*B1_ll + v2_ll*B2_ll + v3_ll*B3_ll
  # extract magnetic field variable from the right and set the normal velocity
  # Note, both depend upon the orientation and need psi_rr
  if mode==1
    if orientation == 1 # x-direction
      v_normal = v1_ll
      B_normal = B1_rr
    else # y-direction
      v_normal = v2_ll
      B_normal = B2_rr
    end
    psi_norm = psi_rr
  elseif mode==2
    if orientation == 1 # x-direction
      v_normal = v1_ll
      B_normal = B1_rr - B1_ll
    else # y-direction
      v_normal = v2_ll
      B_normal = B2_rr - B2_ll
    end
    psi_norm = psi_rr - psi_ll
  else #mode==3
    if orientation == 1 # x-direction
      v_normal = v1_ll
      B_normal =-B1_ll
    else # y-direction
      v_normal = v2_ll
      B_normal =-B2_ll
    end
    psi_norm =-psi_ll
  end

  # compute the nonconservative flux: Powell (with B_normal) and Galilean (with v_normal)
  noncons2 = 0.5 * B_normal * B1_ll
  noncons3 = 0.5 * B_normal * B2_ll
  noncons4 = 0.5 * B_normal * B3_ll
  noncons5 = 0.5 * B_normal * v_dot_B_ll + 0.5 * v_normal * psi_ll * psi_norm
  noncons6 = 0.5 * B_normal * v1_ll
  noncons7 = 0.5 * B_normal * v2_ll
  noncons8 = 0.5 * B_normal * v3_ll
  noncons9 = 0.5 * v_normal * psi_norm

  return SVector(0, noncons2, noncons3, noncons4, noncons5, noncons6, noncons7, noncons8, noncons9)
end
```

Reply:
> I tested the rotor with `@benchmark Trixi.run("examples/2d/parameters_rotor.toml",t_end=0.05) evals=20`

In general, I prefer to benchmark lower-level functions like `rhs!` or `calc_volume_integral!` that take less time, are less influenced by possible noise, and more sensitive the the actual changes you did. For example, a complete run can use IO etc. You could do that as well to ensure that there is no performance difference.

Reply:
Did you push the new implementation? I can't see it on GitHub right now...

Anyway, could you please add some comments to describe `mode`?

Reply:
And maybe use more descriptive types of `mode`, e.g. symbols?

Reply:
I have not pushed yet, as I wanted to hear your opinion first. I only added this comments in the description of the function
```
"""
    noncons_interface_flux(u_left, u_right, orientation, mode, equation::IdealGlmMhdEquations2D)
    
Strong form of nonconservative flux on a side (Powell and GLM terms)
     phi^L 1/2 (B^L+B^R) normal - phi^L B^L normal = phi^L 1/2 (B^R-B^L) normal
**OBS! 1) The non-conservative interface flux depends on the discretization. Following "modes" are available:
        * 1: 'weak' formulation of split DG already includes the contribution -1/2(phi^L B^L normal)
             so this mode only adds 1/2(phi^L B^R normal)
             ... analogously for the Galilean nonconservative term
        * 2: This mode adds the whole non-conservative term: phi^L 1/2 (B^R-B^L)
        * 3: This mode adds the split-form DG volume integral contribution: This is equivalent to (2)-(1)
             - 1/2(phi^L B^L)**
     2) this is non-unique along an interface! normal direction is super important
"""
```

Reply:
The first OBS describes `mode`

Reply:
Thanks! That helps. Could you nevertheless experiment with something more descriptive like symbols, please?

And feel free to push your changes to this PR. It's much easier to discuss them in that way, I think.

Reply:
Following your advice to test only `rhs!`, I used following testing script:
```julia
using Trixi; using BenchmarkTools

# Initialize everything 
parameter_file="path/to/parameter/file.toml";
Trixi.init_parameters(parameter_file);
mesh, solver, time_parameters, time_integration_function = Trixi.init_simulation();

# Compute RHS once to avoid overhead in the benchmark
Trixi.rhs!(solver,0.0)

# Benchmark!
@benchmark Trixi.rhs!(solver,0.0)
```
And got following results for the rotor:
1. New implementation:
```
BenchmarkTools.Trial: 
  memory estimate:  11.53 KiB
  allocs estimate:  83
  --------------
  minimum time:     6.758 ms (0.00% GC)
  median time:      6.792 ms (0.00% GC)
  mean time:        6.829 ms (0.00% GC)
  maximum time:     12.054 ms (0.00% GC)
  --------------
  samples:          733
  evals/sample:     1
```
2. Old implementation:
```
BenchmarkTools.Trial: 
  memory estimate:  11.53 KiB
  allocs estimate:  83
  --------------
  minimum time:     6.670 ms (0.00% GC)
  median time:      6.694 ms (0.00% GC)
  mean time:        6.712 ms (0.00% GC)
  maximum time:     10.046 ms (0.00% GC)
  --------------
  samples:          745
  evals/sample:     1
```

and the Alfvén waves:
1. New implementation
```
BenchmarkTools.Trial: 
  memory estimate:  9.58 KiB
  allocs estimate:  68
  --------------
  minimum time:     1.891 ms (0.00% GC)
  median time:      1.902 ms (0.00% GC)
  mean time:        1.905 ms (0.00% GC)
  maximum time:     2.788 ms (0.00% GC)
  --------------
  samples:          2623
  evals/sample:     1
```
2. Old implementation:
```
BenchmarkTools.Trial: 
  memory estimate:  9.58 KiB
  allocs estimate:  68
  --------------
  minimum time:     1.900 ms (0.00% GC)
  median time:      1.912 ms (0.00% GC)
  mean time:        1.918 ms (0.00% GC)
  maximum time:     3.476 ms (0.00% GC)
  --------------
  samples:          2606
  evals/sample:     1
```

The `maximum time` varies a lot, even in the same implementation. Therefore, I think that we should focus on the other measured times.

It seems that the new implementation slightly affects the performance when the shock-capturing volume integral is computed..

Reply:
I just pushed the new implementation in 32dca4d37a7d18012860fbcdaa2d8865b5634c6c

Reply:
> The `maximum time` varies a lot, even in the same implementation. Therefore, I think that we should focus on the other measured times.

Yeah, the maximum time is mostly meaningless. I usually concentrate on the median/mean results.

> It seems that the new implementation slightly affects the performance when the shock-capturing volume integral is computed.. However, the difference is only ~1-2%.

Did you run the benchmarks multiple times? The mean/median times often fluctuate by a few percent when doing this. On my computer, there isn't a significant performance difference when benchmarking `rhs!`. 

Although it doesn't seem to matter much in this example: When you benchmark in global scope as you do in your script, 
you should always interpolate variables, i.e. use
```julia
@benchmark Trixi.rhs!($(solver), $(0.0))
```

Reply:
I think it's definitely fine to use the new method :+1: A quick test on my system indicates that using `Symbol`s instead of integers for `mode` doesn't reduce the performance. Hence, you could use

1. `mode == :weak` 
2. `mode == :whole`
3. `mode == :inner`

or something similar to make the parameter `mode` a bit more descriptive. I took the names above basically from your previous version.

Reply:
> Although it doesn't seem to matter much in this example: When you benchmark in global scope as you do in your script, you should always interpolate variables, i.e. use
> 
> @benchmark Trixi.rhs!($(solver), $(0.0))

Thanks for the tip! 

> I think it's definitely fine to use the new method +1 A quick test on my system indicates that using Symbols instead of integers for mode doesn't reduce the performance. 

Great. Thanks for testing. I'll add the symbols :)

Reply:
The symbols were added in c7be5cab7dc8613cf5d68af4dede200f8293d77f